spec.bs

```diff
@@ -548,6 +553,11 @@ input {{SecurePaymentConfirmationRequest}} |data|, are:
 1. If |data|["{{SecurePaymentConfirmationRequest/credentialIds}}"] is empty,
     return `false`.
 
+1. Let |parsedRpId| be the result of running the [=URL parser=] on
+    |data|["{{SecurePaymentConfirmationRequest/rpId}}"].
+
+1. If |parsedRpId| is failure, then return `false`.
```

Comment:
Can you explain these steps further? In the description you say "This ensures that this is a breaking change before any UI is shown.", but I don't quite follow what that means.

Reply:
The intention was that if a user (developer) of SPC does not include the rpId input, then SPC fails before any UI is shown, since the rpId is only actually validated during webauthn authentiation after the UI is accepted. Although I now realize this isn't really an issue if it's a required input field, since the SecurePaymentConfirmationRequest creation would fail if rpId isn't included.

This might still be a reasonable preliminary check, e.g. if the developer is dynamically setting rpId and something went wrong and they've set it to a null/empty string, we can fail SPC before UI is shown here. WDYT?

Reply:
> since the SecurePaymentConfirmationRequest creation would fail if rpId isn't included.

Yep, exactly!

> This might still be a reasonable preliminary check, e.g. if the developer is dynamically setting rpId and something went wrong and they've set it to a null/empty string, we can fail SPC before UI is shown here. WDYT?

So I think we should probably check it here, because WebAuthn actually doesn't as currently spec'd (surprise!). WebAuthn isn't checking it for two reasons:

1. We don't set `options.rpId` in step 4 [of this alg](https://w3c.github.io/secure-payment-confirmation/#sctn-steps-to-respond-to-a-payment-request) - this is a bug and we should fix that.
2. We [override](https://w3c.github.io/secure-payment-confirmation/#client-extension-processing-authentication) step 6.1 of [this WebAuthn algorithm](https://w3c.github.io/webauthn/#dom-publickeycredential-discoverfromexternalsource-slot), which removes any check that would happen for `rpId` being a domain.

So I think we should either do that check here, or put the check into our extension override. In either case, we need to do more than just parse the URL - we need to make sure its a specifically [valid domain](https://url.spec.whatwg.org/#valid-domain). I think we can just reference that URL spec algorithm to do so, it seems to take a string as input and check that its a domain?

Reply:
IIUC the rpId is already validated as part of this spec in [Verifying an Authentication Assertion](https://w3c.github.io/secure-payment-confirmation/#sctn-verifying-assertion), step 2.5 _"After step 13, insert the following steps: ... Verify that the value of C["[payment](https://w3c.github.io/secure-payment-confirmation/#dom-collectedclientpaymentdata-payment)"]["[rp](https://w3c.github.io/secure-payment-confirmation/#dom-collectedclientadditionalpaymentdata-rp)"] matches the [Relying Party](https://w3c.github.io/webauthn/#relying-party)'s origin."_. Doesn't verifying that it matches effectively do the domain check? Or should we also make sure it's a valid domain here?

Reply:
Those checks are for the Relying Party to make on their own server, after the assertion has been generated. They're not part of any browser (or authenticator) algorithm.

Reply:
Also, I said above 'make the check here or in the authenticator algorithm override', but of course we want it to happen before any UI is shown so it should go here.

Reply:
> Done. So IIUC this means webauthn will now validate that rpId matches the credential's RP.

Yep!

> Done, although #valid-domain isn't something that's exported from the url spec, so I have to link to it directly, since [[url#valid-domain]] doesn't work. Do you think that's okay to do?

See my comment on that line.